aip/aog/3021.md

```diff
@@ -0,0 +1,181 @@
+---
+aip:
+  id: 3021
+  state: draft
+  created: 2019-10-04
+  updated: 2019-10-04
+  type: guidance
+js:
+  - /assets/js/graphviz/viz.js
+  - /assets/js/graphviz/lite.render.js
+  - /assets/js/aip/aip-graphviz.js
+permalink: /3021
+---
+
+# Type protos in Actions on Google APIs
+
+Many Actions on Google fulfillment APIs need to represent typed data (datetime,
+money, etc.). In some cases, there are pre-existing common components for
+representing these types.
+
+## Guidance
+
+Fulfillment APIs **must** use the AoG common types when applicable, and **must
+not** define alternative type protos. For example, an API which needs to
+represent a monetary amount **must** use [google.type.Money][money], and **must
+not** define an alternative representation.
+
+APIs **may** wrap existing types in new protos. This may be useful in order to
+define compound types, provide additional metadata, provide semantic signals
+(if the typed value was obtained from a user query), etc.
+
+## Actions on Google common types
+
+The AoG common types include all the protobuf types and API types defined in
+[AIP-213][], as well as all the protos in
+[google.actions.type.\*][actions-type].
+
+### Adding new types
+
+If a given API needs a type for which there is no existing common type, and the
+type is not conceptually specific to that API, and the type is well understood
+enough to be modeled in a permanent/unversioned way, then API author **should**
+consider proposing a new AoG common type in
+[google.actions.type.\*][actions-type].
+
+The guidance in the appendix of [AIP-213][] also applies to creating new AoG
+common types. Specifically, API authors proposing new types **should** consider
+whether a new type should be a Google-wide common type rather than an AoG
+common type. One exception to this is that new AoG common types are not
+proposed by opening a GitHub ticket. Instead, AoG API authors should see
+//google/actions/type/OWNERS for contact information.
+
+## Built-In Intent arguments and common types
```

Comment:
This section is confusing to me. It's not clear what needs to be done by the API author vs the implementation. Maybe taking a simple example all the way through would be helpful. What is the BII? What is the corresponding API? How can implementations do the mapping? In Quick Actions? In a completely custom vertical?

Reply:
Thanks for the feedback - this is a hard section to write.

I thought about suggesting a standard field name (a la the `page_size` field recommended in https://aip.dev/158), but where one API might use `intent_parameter_path` because it has "parameters", an API which has "slots" might want a field named `intent_slot_path` instead.  Maybe all those APIs should migrate (if necessary) to a more generic/API-agnostic field name like `bii_mapping_path`?  But I'm not sure to what extent this really addresses your question.

I think a more thorough example would be a good idea.  I'm not sure if it's permissible to do that with APIs that aren't public yet, though.  I suppose I could make up a toy API for the sake of illustration.  Do you have an opinion on that vs. a more realistic example with a real API (possibly blocking this (part of the) AIP until such an API is publicly released)?

> It's not clear what needs to be done by the API author vs the implementation.

This is a fair point, but some other comparable AIPs (e.g. AIP-158) also provide guidance about the API behavior, rather than just the API protos themselves.  For example, "The API may return fewer results than the number requested (including zero results), even if not at the end of the collection."

Do you think that's a fair comparison?  (Genuine question, not 100% convinced myself.)

Reply:
Also, let me know if you want me to dump this in a Google doc for easier commenting.